configuration/configmgr/configmgrlib/mod_templates/EnvModTemplate.cpp

```diff
@@ -0,0 +1,341 @@
+            msg += "Invalid document: " + std::string(sb.GetString());
+            TemplateException te(msg);
+            te.setInvalidTemplate(true);
+            throw te;
+        }
+    }
+
+    //
+    // Now, parse the template, all as part of loading it
+    parseTemplate();
+}
+
+
+void EnvModTemplate::parseTemplate()
+{
+    //
+    // Process the defined template sections. If any errors are found, thrown an exception
+
+    //
+    // Inputs
+    if (m_pTemplate->HasMember("inputs"))
+    {
+        const rj::Value &inputs = (*m_pTemplate)["inputs"];
+        parseInputs(inputs);
+    }
+
+    //
+    // Operations (a required section)
+    parseOperations((*m_pTemplate)["operations"]);
+}
+
+
+void EnvModTemplate::parseInputs(const rj::Value &inputs)
+{
+    for (auto &inputDef: inputs.GetArray())
+    {
+        parseInput(inputDef);
+    }
+}
+
+
+void EnvModTemplate::parseInput(const rj::Value &input)
+{
+    rj::Value::ConstMemberIterator it;
+    std::shared_ptr<Input> pInput;
+
+    //
+    // input name, make sure not a duplicate
+    std::string inputName(input.FindMember("name")->value.GetString());
+    std::string type(input.FindMember("type")->value.GetString());
+    pInput = inputValueFactory(type, inputName);
+
+    //
+    // Get and set the rest of the input values
+    it = input.FindMember("prompt");
+    if (it != input.MemberEnd())
+        pInput->setUserPrompt(it->value.GetString());
+    else
+        pInput->setUserPrompt("Input value for " + inputName);
+
+    it = input.FindMember("description");
+    if (it != input.MemberEnd()) pInput->setDescription(it->value.GetString());
+
+    it = input.FindMember("tooltip");
+    if (it != input.MemberEnd()) pInput->setTooltip(it->value.GetString());
+
+    it = input.FindMember("value");
+    if (it != input.MemberEnd()) pInput->setValue(trim(it->value.GetString()));
+
+    it = input.FindMember("prepared_value");
+    if (it != input.MemberEnd()) pInput->setPreparedValue(trim(it->value.GetString()));
+
+    m_inputs.add(pInput);
+}
+
+
+std::shared_ptr<Input> EnvModTemplate::getInput(const std::string &name, bool throwIfNotFound) const
+{
+    return m_inputs.getInput(name, throwIfNotFound);
+}
+
+
+std::vector<std::shared_ptr<Input>> EnvModTemplate::getInputs(int phase) const
+{
+    return m_inputs.all(phase);
+}
+
+
+void EnvModTemplate::parseOperations(const rj::Value &operations)
+{
+    for (auto &opDef: operations.GetArray())
+    {
+        parseOperation(opDef);
+    }
+}
+
+
+void EnvModTemplate::parseOperation(const rj::Value &operation)
+{
+    std::shared_ptr<Operation> pOp;
+    std::string action(operation.FindMember("action")->value.GetString());
+
+    if (action == "create_node" || action == "create_nodes")
+    {
+        pOp = parseCreateNodeOperation(operation);
+    }
+    else if (action == "noop")
+    {
+        pOp = parseOperationNoop(operation);
+    }
+    m_operations.emplace_back(pOp);
```

Comment:
You are correct, however I hadn't completed all of the work for this. the modify_node operation was to be added later. The goal was to "know" that the operation was valid based on the schema. However, we all get things out of sync, so I added a throw just in case.